aexpect/session_ops.py

```diff
@@ -0,0 +1,598 @@
+# This Python file uses the following encoding: utf-8
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+
+"""
+
+SYNOPSIS
+------------------------------------------------------
+A list of frequent operations performed through remote sessions.
+
+Copyright: 2018 Intra2net AG
+
+
+DESCRIPTION
+------------------------------------------------------
+Most commands passed to a session are routine invocations of common userland
+tools. Nevertheless some of them require a certain effort to set up because
+they are squeezed through the eye of the needle, the shell session.
+
+The functionality in this module assumes a guest with the ordinary userland
+tools available. Some of it might just work on Windows guests if the proper
+environment (i. e. Cygwin) is present.
+
+This module aims to provide convenient wrappers for butter-and-bread
+functionality on a Linux system. Currently, the implemented functions fall
+under four categories: ``grep``, ``stat``, ``test``, and common file ops.
+
+grep:
+Standard case search operations. It is always assumed that binaries in the path
+should be searched (the ``-a`` flag).
+
+stat:
+Allows passing an optional format argument (``-c``). Custom wrappers for
+extracting *atime*, *ctime*, and *mtime*, etc exist.
+
+test:
+Allows for testing existence, permissions and other properties of files.
+
+file ops:
+Simple file operations executed through sessions. Most functions boil down to
+executing simple commands like ls, tar, md5sum, etc. If these operations return
+non-0, a :py:class:`RuntimeError` is raised, containing the commands's error
+message. All functions :py:func:`shlex.quote` their args for better security.
+
+
+IMPLEMENTATION
+------------------------------------------------------
+
+"""
+
+import sys
+import re
+from os.path import join, split as path_split
+
+if sys.version_info[0] < 3:
+    from six.moves import shlex_quote as quote
+else:
+    from shlex import quote   # pylint: disable=E0611
+
+# Need this import for sphinx and other documentation to produce links later on
+# from .client import ShellSession
+
+
+###############################################################################
+# grep(1)
+###############################################################################
+
+GREP_FLAGS_KNOWN = ["a", "r", "v"]
+
+
+def grep(session, expr, path, flags=None):
+    """
+    Invoke ``grep`` on guest searching for *expr* in *path*. Throws a
```

Comment:
Here and most of docstrings are not following PEP-257.

Reply:
Do you mind pointing me to specific conventions where this has a problem? PEP-257 contains multiple such and I am not sure which one you are referring to.

Reply:
Sure, my bad. If you are looking to a particular section, I would recommend this:

https://www.python.org/dev/peps/pep-0257/#id17

But I think the whole PEP-257 would be worthed to read, since it is small and very stratight forward.

Reply:
I have read the PEP-257 before and following the link just redirects me to a section (at least what I see when I click on it). But I think I am starting to answer my question - I guess you mean the first line being separated with the rest. Good catch! The docstring is quite old from before we put started accommodating docstring standards in our code base so I totally glimpsed over it. Thanks!

Reply:
I corrected the `grep` for the initial statement separation. Other than this we already have utilities like `aexpect/remote_door.py` that use a similar style of docstring formatting. Perhaps I could go over all of them in a single PR if you have specific pointers about these docstrings later on.